File: leetcode_python/threesum.js
var threeSum = function(nums) {
    var res = [];
    if(!nums || nums.length < 3) return res;
    nums = nums.sort(function(x,y){return x-y});
    console.log(nums);
    for(var i = 0; i < nums.length-2; i++){
        if(i !== 0 && nums[i] === nums[i-1]){
            continue;
        }
        var l = i+1;
        var r = nums.length-1;
        while(l < r){
            if(nums[l]+nums[r]+nums[i] === 0){
                var tmp = [];
                tmp.push(nums[i]);
                tmp.push(nums[l]);
                tmp.push(nums[r]);
                res.push(tmp);
                l++;
                r--;
                while(l < r && nums[l] === nums[l-1]){
                    l++;
                }
                while(l < r && nums[r] === nums[r+1]){
                    r--;
                }
            }else if(nums[l]+nums[r]+nums[i] < 0){
                l++;
            }else{
                r--;
            }
        }
    }
    return res;
    
};

var res = threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]);
console.log(res);

File: leetcode_python/threeSum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        length = len(nums)
        for i in range(0,length-2):
            if i and nums[i] == nums[i-1]:
                continue
            target = nums[i]
            left,right = i+1,length-1
            while left < right:
                if nums[left]+nums[right]+target == 0:
                    res.append([nums[i],nums[left],nums[right]])
                    right-=1
                    left+=1
                    while left < right and nums[left] == nums[left-1]:
                        left += 1
                    while left < right and nums[right] == nums[right+1]:
                        right -= 1
                elif nums[left]+nums[right]+target < 0:
                    left+=1
                else:
                    right-=1
        return res
        